enhancements/machine-api/vsphere-data-disk.md

```diff
@@ -0,0 +1,518 @@
+// +kubebuilder:validation:Enum=Thin;Thick;EagerlyZeroed
+type ProvisioningMode string
+
+const (
+    // ProvisioningModeThin creates the disk using thin provisioning. This means a sparse (allocate on demand)
+    // format with additional space optimizations.
+    ProvisioningModeThin ProvisioningMode = "Thin"
+
+    // ProvisioningModeThick creates the disk with all space allocated.
+    ProvisioningModeThick ProvisioningMode = "Thick"
+
+    // ProvisioningModeEagerlyZeroed creates the disk using eager zero provisioning. An eager zeroed thick disk
+    // has all space allocated and wiped clean of any previous contents on the physical media at
+    // creation time. Such disks may take longer time during creation compared to other disk formats.
+    ProvisioningModeEagerlyZeroed ProvisioningMode = "EagerlyZeroed"
+)
+```
+
+Examples of the various CRs using the above type enhancement:
+
+CPMS
+```yaml
+apiVersion: machine.openshift.io/v1
+kind: ControlPlaneMachineSet
+metadata:
+  creationTimestamp: "2024-11-06T15:30:46Z"
+  finalizers:
+  - controlplanemachineset.machine.openshift.io
+  generation: 1
+  labels:
+    machine.openshift.io/cluster-api-cluster: ngirard-multi-rtmvw
+  name: cluster
+  namespace: openshift-machine-api
+  resourceVersion: "17330"
+  uid: fe7f5e45-40da-4c69-b1b9-5cc415970af4
+spec:
+  replicas: 3
+  selector:
+    matchLabels:
+      machine.openshift.io/cluster-api-cluster: ngirard-multi-rtmvw
+      machine.openshift.io/cluster-api-machine-role: master
+      machine.openshift.io/cluster-api-machine-type: master
+  state: Active
+  strategy:
+    type: RollingUpdate
+  template:
+    machineType: machines_v1beta1_machine_openshift_io
+    machines_v1beta1_machine_openshift_io:
+      failureDomains:
+        platform: VSphere
+        vsphere:
+        - name: fd-1
+      metadata:
+        labels:
+          machine.openshift.io/cluster-api-cluster: ngirard-multi-rtmvw
+          machine.openshift.io/cluster-api-machine-role: master
+          machine.openshift.io/cluster-api-machine-type: master
+      spec:
+        lifecycleHooks: {}
+        metadata: {}
+        providerSpec:
+          value:
+            apiVersion: machine.openshift.io/v1beta1
+            credentialsSecret:
+              name: vsphere-cloud-credentials
+            dataDisks:
+            - diskSizeGiB: 10
+              name: "etcd"
+              provisioningMode: "Thick"
+            - diskSizeGiB: 20
+              name: "container-images"
+              provisioningMode: "Thin"
+            diskGiB: 100
+            kind: VSphereMachineProviderSpec
+            memoryMiB: 16384
+            metadata:
+              creationTimestamp: null
+            network:
+              devices: null
+            numCPUs: 8
+            numCoresPerSocket: 2
+            snapshot: ""
+            template: ""
+            userDataSecret:
+              name: master-user-data
+            workspace: {}
```

Comment:
have we done any testing with multiple failure domains and additional data disks? We have some webhook/validation logic that I don't think would be impacted by this, but I just want to make sure.

Reply:
I have tested creating vms with multiple disks in one of the FDs in a multi FD cluster.  I can run a quick test where i spin up several across all FDs to validate this. These disks are created as part of the vm clone process.  I'd expect the webhook not to care much other than validating the vlaues of each disk config, but I'll rerun and confirm.

Reply:
Yep, it is all looking good.